tiled/server/router.py

```diff
@@ -67,357 +66,436 @@
-        expires=datetime.now(timezone.utc) + timedelta(seconds=600),
-    )
-
 
-async def search(
-    request: Request,
-    path: str,
-    fields: Optional[List[schemas.EntryFields]] = Query(list(schemas.EntryFields)),
-    select_metadata: Optional[str] = Query(None),
-    offset: Optional[int] = Query(0, alias="page[offset]", ge=0),
-    limit: Optional[int] = Query(
-        DEFAULT_PAGE_SIZE, alias="page[limit]", ge=0, le=MAX_PAGE_SIZE
-    ),
-    sort: Optional[str] = Query(None),
-    max_depth: Optional[int] = Query(None, ge=0, le=DEPTH_LIMIT),
-    omit_links: bool = Query(False),
-    include_data_sources: bool = Query(False),
-    entry: Any = SecureEntry(scopes=["read:metadata"]),
-    query_registry=Depends(get_query_registry),
-    principal: str = Depends(get_current_principal),
-    **filters,
-):
-    request.state.endpoint = "search"
-    if entry.structure_family != StructureFamily.container:
-        raise WrongTypeForRoute("This is not a Node; it cannot be searched or listed.")
-    try:
-        resource, metadata_stale_at, must_revalidate = await construct_entries_response(
-            query_registry,
-            entry,
-            "/search",
-            path,
-            offset,
-            limit,
-            fields,
-            select_metadata,
-            omit_links,
-            include_data_sources,
-            filters,
-            sort,
-            get_base_url(request),
-            resolve_media_type(request),
-            max_depth=max_depth,
-        )
-        # We only get one Expires header, so if different parts
-        # of this response become stale at different times, we
-        # cite the earliest one.
-        entries_stale_at = getattr(entry, "entries_stale_at", None)
-        headers = {}
-        if (metadata_stale_at is None) or (entries_stale_at is None):
-            expires = None
-        else:
-            expires = min(metadata_stale_at, entries_stale_at)
-        if must_revalidate:
-            headers["Cache-Control"] = "must-revalidate"
-        return json_or_msgpack(
-            request,
-            resource.model_dump(),
-            expires=expires,
-            headers=headers,
-        )
-    except NoEntry:
-        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail="No such entry.")
-    except WrongTypeForRoute as err:
-        raise HTTPException(status_code=HTTP_404_NOT_FOUND, detail=err.args[0])
-    except JMESPathError as err:
-        raise HTTPException(
-            status_code=HTTP_400_BAD_REQUEST,
-            detail=f"Malformed 'select_metadata' parameter raised JMESPathError: {err}",
-        )
+def get_router(
+    query_registry: QueryRegistry,
+    serialization_registry: SerializationRegistry,
+    deserialization_registry: SerializationRegistry,
+    validation_registry: ValidationRegistry,
+) -> APIRouter:
+    router = APIRouter()
+
+    T = TypeVar("T")
+
+    def patch_route_signature(route: Callable[..., T]) -> Callable[..., T]:
```

Comment:
As written here, this closes over `query_registry`. Another option would be to define this function at global scope and make it a higher-order decorator that takes `query_registry` as an argument.

Either is fine with me, not sure which I prefer, just considering options out loud.